### ISSUE TITLE
Terminology review.

### ISSUE THREAD

**Issue by exerro** [Exerro/amend#15]
The terminology is getting to be a bit of a mess. It has to fit in with the flowchart shown below:
https://drive.google.com/file/d/0B37gxkuD6apmS3lvLXFFUHRQbnM/view?usp=sharing
![](https://i.imgur.com/NpELWC4.png)

* `Executable` - A file that can be run in normal Lua.
* `Package` - A set of independent files functioning together in some way.
* `Plugin` - A set of procedures that alter and extend the normal build process, adding in new functionality, or changing existing functionality.

So, just to list everything on the flow chart:
* `Protocol` - A reference to a procedure.
* `Input file name` - A string identifier used to include files in a build. It is composed of an optional protocol (e.g. `http`, `github`, `file`), colon, and the file identifier. Protocol defaults to `file`.
* `File URI` - A reference to a file, composed of a protocol (e.g. `http`, `file`), delimiter (`://`), and the absolute path to the file (e.g. `github.com/Exerro/amend/`, `dir/file.lua`).
* `Name resolver` - A process in which input file names are mapped to a list of File URIs.
* `Stream` - An object used to pass through the contents of a file (allowing the file not to be entirely read).
* `Stream generator` - A process in which File URIs are used to create a stream for each URI.
* `AST` (Abstract syntax tree) - An intermediate represenation of the grammar of a language.
* `Token` - An object representing a distinct section of text in a stream, e.g. a word, number, or string. Also contains position information.
* `Lexer` - An object generating tokens from a stream.
* `Parser` - An object using a stream to generate an AST. Internally uses a lexer.
* `AST modification` - A process in which an AST is modified (resulting in another AST) by various callbacks, applied iteratively over the entire AST, or particular nodes.
* `AST merge` - A process in which multiple ASTs from multiple files are merged into a single AST.
* `Compiled output` - The result of a build, e.g. the executable, or a package.

There are probably more things, comment and I'll add them.

**Comment by viluon** [comment]
I think you documented the less confusing parts of the system. We need to decide on the build/task/pipeline/plugin part...

**Comment by exerro** [comment]
Yep. I was in a rush and now I'm stuck with the patchiest mobile data signal that only works when I'm outside, away from my laptop which is on 7% battery. Life is cruel sometimes. Basically a build is an independent operation on a group of files, a pipeline is probably badly named and not really necessary so let's get rid of that, a plugin exposes callbacks for all the various processes, and you should work on tasks
